Add Yiddish Braille Table

abbyhowell opened the pull request:
I'm a Yiddish teacher (and software engineer), and I recently started teaching Yiddish to a blind student who uses Braille. We're working together to figure out how to make Yiddish texts accessible to him. Yiddish is written in the Hebrew alphabet, so it seemed initially to me that Hebrew Braille should be able to handle it, but what we found out is that there are a handful of unicode characters that are common in Yiddish which have their own unicode representations, and do not map to any letters in the Hebrew Braille table.

This PR adds an uncontracted braille table for Yiddish, based on the Hebrew braille table. Yiddish was incorporated into unicode two decades ago, and Yiddish unicode uses several variations on letters and combinations of letters which have unique unicode characters, which are not represented by the Hebrew Braille table. Which means that, when a blind reader tries to read a Yiddish text file, many of the characters are not represented. My hope with this pull request is to provide a first attempt to fill in the blanks so that braille reader software will be able to recognize the unicode characters that represent Yiddish and display them in readable Braille.

After a semester of reading Yiddish in braille using a series of variations on this table, my student and I, in consultation with a consortium of Yiddish-speaking software engineers and multilingual Braille users, have finalized our proposal for how to represent Yiddish in Braille.

egli reviewed (changes requested):
This is fascinating!

I know very little about Yiddish and just read the Wikipedia page. 

First off: Your PR looks good. It just needs a few amendments :-)
- [ ] You only need one file. Just merge the contents of `yi.tbl` into `yi.utb`
- [ ] Your description (of the PR) is very good. Since there doesn't seem to be an official braille standard for Yiddish (isn't there?) this documentation should probably be included in the table itself
- [ ]  Can you provide a small YAML test file that tests a few Yiddish sentences and maybe some corner cases. Maybe you can look at `he-IL.yaml` for inspiration
- [ ] You are saying that the table is based off the Hebrew table but as far as I can tell you do not include the Hebrew table. Is there some duplication? Could you maybe include the Hebrew table and overwrite what is different?

Thanks for this excellent PR and I hope we can include it in the June release

egli commented:
Hi @abbyhowell I integrated the table in the build system so that the YAML tests are actually run. They fail and it seems to be related to the letter alef. Can you have a look at the [test result](https://github.com/liblouis/liblouis/actions/runs/8277823432/artifacts/1325255545)?

abbyhowell commented:
Thank you so much @egli ! I am so glad you encouraged me to write tests, because I had typos in my braille table that I hadn't realized until I saw the test failures. I've gone back and updated the tests, and I've rewritten the braille table to only include the unicode characters which are not in the Hebrew table, and I've updated the braille mappings to render simply the underlying Hebrew characters. It seems like it will be the responsibility of a future contracted braille table to contract these special characters into a single braille character.

markhdavid commented:
Hi @abbyhowell, I'm not at all familiar with computerized braille but very familiar with Yiddish and Unicode. The thing is, Yiddish and Hebrew are *not* encoded as separate languages. They are completely overlapping. Unicode does not even encode languages, just scripts. There are single code points for YIVO standard combinations, but they're there for historical compatability purposes and should be avoided in modern software if pissibke. What is the reason to use precombined characters vs composed characters?

markhdavid commented:
The Unicode character combinations needed for Yiddish (Yivo orthography) are shown here:

[https://uyip.org/combinations/](https://uyip.org/combinations/)

markhdavid commented:
In Unicode both Hebrew-language and Yiddish-language text can be be encoded with code points in the main Hebrew block. Assuming there's a system for handling Hebrew-language text, including nikud, generally, there are three main issues with this that concern Yiddish: (1) dealing with precomposed characters in the Alphabetic Presentation Forms section of Unicode; (2) dealing with the 3 so-called Yiddish digraph characters in the main Hebrew block, which generally are not used for Hebrew language text; and (3) dealing the diacritic RAFE, which generally is not used for Hebrew language text.

Here are basic approaches to dealing with them:

For (1) you can apply Unicode Normalization Form D (NFD). In NFD, fir the Hebrew script, characters are decomposed to their constituent components, often resulting in 2 code points (in the main Hebrew blick) representing what visually appears to be a single character. Note that U+FB1F (= *pasekh-tsvey-yudn*) decomposes to `Unicode Character 'HEBREW LIGATURE YIDDISH DOUBLE YOD' (U+05F2)` and `Unicode Character 'HEBREW POINT PATAH' (U+05B7)`, which *are* both in the main Hebrew block.

For (2) and (3) it's best for there to be representations for these characters. Regarding (2), it's particularly important to represent tsvey-yudn and tsvey-vovn as distinct from a series of 2 yuds or 2 vovs, especially for older orthographies, since they can combine with a variety of nikud. The combination in *pasekh-tsvey-yudn* is best known, since it's used in the modern standard Yivo orthography. Regarding (3) the RAFE diacritic is no longer used for modern Hebrew with nikud, so it's often excluded/forgotten in systems devised for the Hebrew language.

Here is where I could use clarification. Are there Braille representations for the following 4 characters?

> 05F0;HEBREW LIGATURE YIDDISH DOUBLE VAV
> 05F1;HEBREW LIGATURE YIDDISH VAV YOD
> 05F2;HEBREW LIGATURE YIDDISH DOUBLE YOD
> 05BF;HEBREW POINT RAFE

abbyhowell commented:
> In Unicode both Hebrew-language and Yiddish-language text can be be encoded with code points in the main Hebrew block. Assuming there's a system for handling Hebrew-language text, including nikud, generally, there are three main issues with this that concern Yiddish: (1) dealing with precomposed characters in the Alphabetic Presentation Forms section of Unicode; (2) dealing with the 3 so-called Yiddish digraph characters in the main Hebrew block, which generally are not used for Hebrew language text; and (3) dealing the diacritic RAFE, which generally is not used for Hebrew language text.
> 
> Here are basic approaches to dealing with them:
> 
> For (1) you can apply Unicode Normalization Form D (NFD). In NFD, fir the Hebrew script, characters are decomposed to their constituent components, often resulting in 2 code points (in the main Hebrew blick) representing what visually appears to be a single character. Note that U+FB1F (= _pasekh-tsvey-yudn_) decomposes to `Unicode Character 'HEBREW LIGATURE YIDDISH DOUBLE YOD' (U+05F2)` and `Unicode Character 'HEBREW POINT PATAH' (U+05B7)`, which _are_ both in the main Hebrew block.
> 
> For (2) and (3) it's best for there to be representations for these characters. Regarding (2), it's particularly important to represent tsvey-yudn and tsvey-vovn as distinct from a series of 2 yuds or 2 vovs, especially for older orthographies, since they can combine with a variety of nikud. The combination in _pasekh-tsvey-yudn_ is best known, since it's used in the modern standard Yivo orthography. Regarding (3) the RAFE diacritic is no longer used for modern Hebrew with nikud, so it's often excluded/forgotten in systems devised for the Hebrew language.
> 
> Here is where I could use clarification. Are there Braille representations for the following 4 characters?
> 
> > 05F0;HEBREW LIGATURE YIDDISH DOUBLE VAV
> > 05F1;HEBREW LIGATURE YIDDISH VAV YOD
> > 05F2;HEBREW LIGATURE YIDDISH DOUBLE YOD
> > 05BF;HEBREW POINT RAFE

@markhdavid I'm so happy you came over to advise! Thank you so much for coming over here to help with this! 

There are not currently Braille representations for those 4 characters. I had included the first three in my proposed table but I had entirely overlooked the Rafe. I will have to figure out how to put composed unicode characters into the braille table, and so I can figure out what to do with the Rafe.

(note to non-Yiddish-speaking readers: a Rafe is an accent mark that looks like a horizontal line over the letter. Just like in English, the letter C can be pronounced like a K or like an S, Hebrew has several letters that can be pronounced in more than one way, and sometimes they use extra accent markers to indicate which pronunciation is meant. The Rafe is one of those accent markers, and it is sometimes, but not consistently, used in Yiddish to indicate the difference between pey פּ  and fey פֿ, or between beys ב and veys בֿ). 

Before this semester, I didn't know much of anything about how Yiddish is represented in Unicode, and I didn't know about the difference between precombined vs composed letters until now, so I'm very glad you stopped in to help. My process so far has been to have my Yiddish student load up a Yiddish text file in his braille computer and then read it to me, and make a mental list of the characters that he reads out as "mystery character" - which were mostly the digraph characters, but have included khof כ and fey פֿ (although I didn't, until now, understand why that might be.)

My goal, I think, is to find all the possible characters that could show up in a Yiddish text and to make sure that, when they get read into Braille, that what comes out on the other end reads like it's written in the Hebrew alphabet. 

Harry Bochner pointed out to me that because the vowels and diacritics are often omitted, and because a text could have tsvey-vovn as two vovs or as a digraph character, that it wouldn't be good to treat the digraph tsvey-vovn differently from two vovs. And I'm not familiar enough with how LibLouis works to know whether I could use a braille table to teach it to decompose a tsvey-vovn character into two vovs, and then to combine them into a tsvey-vov contraction.

My student, Matthew Shifrin, says that he strongly prefers contracted braille because it takes up less space. He audibly cringed when I suggested rendering pasekh-tsvey-yudn as yud-yud-pasekh, but Harry Bochner pointed out that pasekh-tsvey-yudn could show up in a text as yud-yud, or digraph tsvey-yudn, or as pasekh-tsvey-yudn, and that it would be far less confusing if all of those rendered in an essentially similar way. Matthew suggested that if tsvey-yudn is ⠚⠚245-245, then pasekh-tsvey-yudn could be the same shape, but lower down: ⠴⠴ 356-356. So I think I'm going to propose that.

And then I started thinking (and I haven't run this by Matthew yet) - what if we used 8-dot braille, and put the pasekhs and kometses and rafes in the fourth row of the braille table? That way they would be there to provide extra information, or you could slice off the fourth row for six-dot braille and have basically the same amount of information as you would if you were reading a Yiddish text without the diacritics.

bertfrees commented:
Note that we have a table metadata tag `#-unicode-form` that can be used to indicate which Unicode normalization form the input is expected to be in.

It might also make sense to support multiple forms. The table could, in a first stage, using `correct` rules, normalize the input to a specific form, and then map the normalized characters to braille. See also https://github.com/liblouis/liblouis/issues/98.

abbyhowell commented on tables/he-IL.utb:
```diff
@@ -110,8 +110,8 @@ word 'twas =
-word כּ 13 # Hebrew letter PE with dagesh as a separate dot.
-word פּ 1234 # Hebrew letter KAF with dagesh as a separate dot.
+word כּ 13 # Hebrew letter KAF with dagesh as a separate dot.
+word פּ 1234 # Hebrew letter PE with dagesh as a separate dot.
```
this looks like it was a typo in the comment, so I thought I should fix it while I was in there.

markhdavid commented:
> Harry Bochner pointed out to me that because the vowels and diacritics are often omitted, and because a text could have tsvey-vovn as two vovs or as a digraph character, that it wouldn't be good to treat the digraph tsvey-vovn differently from two vovs. And I'm not familiar enough with how LibLouis works to know whether I could use a braille table to teach it to decompose a tsvey-vovn character into two vovs, and then to combine them into a tsvey-vov contraction.

I think that's OK as far as *tsvey-vovn* or *tsvey-yudn* as base characters *without* nikud, and perhaps/hopefully Harry did not mean it in the case where the digraph character is followed by a nikud character. But it's not OK with nikud.  In the Hebrew script, ײַ is *ay*. But ייַ is something different: something like *y&#601; ya*. And יַי is something different again: something like *yay*.  Similarly, װַ is *va*, but ווַ is something else: something like *v&#601; va*. And וַו is something different again: something like *vav*.

> My student, Matthew Shifrin, says that he strongly prefers contracted braille because it takes up less space. He audibly cringed when I suggested rendering pasekh-tsvey-yudn as yud-yud-pasekh, but Harry Bochner pointed out that pasekh-tsvey-yudn could show up in a text as yud-yud, or digraph tsvey-yudn, or as pasekh-tsvey-yudn, and that it would be far less confusing if all of those rendered in an essentially similar way.

I don't think I understand "contracted braille" and how it relates, but I also cringe at *yud-yud-pasekh*. It's true that something might be written in an orthography that does *not* distinguish יי *ey* from ײַ *ay*, writing them both as a pair of yuds. If so, and they encode two yuds as ײ (unicode `U+05F2`), then go ahead and change that to two yuds. But if they do distinguish, then I think it's best not to throw away the distinction *and* best not to turn it into something different like ייַ (= *yud-yud-pasekh*). By the way, there are orthographies that write *ey* as ײֵ (*tseyre-tsvey-yudn*), used, for example, in Harkavy's *[Yidish-English-Hebreyisher Verterbukh](https://www.yiddishbookcenter.org/collections/yiddish-books/spb-nybc211297/harkavy-alexander-yidish-english-hebreyisher-verterbukh)* (1928).

> Matthew suggested that if tsvey-yudn is ⠚⠚245-245, then pasekh-tsvey-yudn could be the same shape, but lower down: ⠴⠴ 356-356. So I think I'm going to propose that.
> And then I started thinking (and I haven't run this by Matthew yet) - what if we used 8-dot braille, and put the pasekhs and kometses and rafes in the fourth row of the braille table? That way they would be there to provide extra information, or you could slice off the fourth row for six-dot braille and have basically the same amount of information as you would if you were reading a Yiddish text without the diacritics.

Sorry, not up to speed enough to understand.

abbyhowell commented:
Hi @egli I am going to mark this as draft and work on it some more. I asked around and I got some good advice from some Yiddish experts and some more good advice from some Braille experts, and I am going to go back to the drawing board and come back to you with a new proposal.

abbyhowell commented:
ok @egli I figured out how to run the tests locally and I got my tests passing (!!) and I got an ad-hoc advisory board of Yiddishists to check my work. So I think this is ready for review now, finally. No rush, obvs.

abbyhowell commented:
thank you for running the tests for me! I thought I had them passing, but I apparently did not. My overrides of the Hebrew table are not overriding as I expected, and I didn't realize it because it turns out that, although I had installed `libyaml` with homebrew, the application didn't find it. I had to run: 

```
export CPPFLAGS="-I/opt/homebrew/Cellar/libyaml/0.2.5/include $CPPFLAGS"
export LDFLAGS="-L/opt/homebrew/Cellar/libyaml/0.2.5/lib $LDFLAGS"
./configure
make
sudo make install
make check
```
and that reproduced the test failure for me locally. 

I'm working on figuring out how to override definitions in the included hebrew table. thank you so much for your help and patience as I figure this out!

abbyhowell commented:
Ok @egli thank you so much for sticking with me here. I finally figured out how to compile the program and run the tests locally, and that was very helpful. I think it's actually working now. Please let me know what you think! 

I ended up using the `always` opcode to override the Hebrew Braille table's instinct to write the "nikudim" diacritic markers separately from the letters when they are part of a multi-part composed unicode character. In Hebrew, the "nikudim" (the dots and lines that are added to letters to indicate vowels) are optional, so it makes sense that they write them separately when they do show up. But in Yiddish, they are just a regular part of some of the letters of the alphabet (פֿ for F, פּ for P, אַ for A, אָ for O, etc), so I wanted them to appear as single characters in Yiddish Braille as well.

abbyhowell commented:
Well, I thought I had this all ready to go, and then I heard back from Caryn Navy at Duxbury Braille Systems, and she has developed a slightly different Yiddish Braille table. So it seems like we have to come to an agreement on what the table should be. Apologies again for the back and forth. I'm going to convene a consortium of Yiddish Braille people, and I'll fix up this table to represent the decision that gets made.

abbyhowell commented:
Hi @egli ! 

My student and I have gotten together with Caryn Navy from Duxbury systems (and a consortium of other interested experts) and have worked out a plan for how to represent Yiddish in Braille. In the end, after a semester of learning and relearning to read Yiddish using different versions of this Braille table created by me and by Caryn, my student Matthew himself dictated to me the final mapping of Yiddish characters to Braille characters that makes the most sense. The final mapping overrides some of the Hebrew Braille characters, but we believe that's appropriate, because Yiddish itself maps sounds to letters differently than Hebrew does. 

Please let me know if there's anything else you'd like me to change. I got the tests passing locally and I've updated the comments in the PR to explain how we got to this point and what decisions we made. 

Thank you so much for your help!

bertfrees commented on tables/yi.utb:
```diff
@@ -0,0 +1,234 @@
+# License along with liblouis. If not, see
+# <http://www.gnu.org/licenses/>.
+
+#-index-name: Yiddish, uncontracted
+#-display-name: Yiddish uncontracted braille
```
I suggest dropping the "uncontracted". This is what we do with all "uncontracted" tables that don't have a "contracted" table in Liblouis. Are there contractions in Yiddish braille?

bertfrees commented on tests/braille-specs/yi.yaml:
```diff
@@ -0,0 +1,72 @@
+# Copyright © 2024 by BAUM Engineering <liblouis@baum.ro>
```
Should you add yourself here?

bertfrees reviewed (changes requested):
Sorry for the late reaction, only reviewed the table today. It turned out to be a very nice table!

I only have these minor remarks:
- see comments in the code regarding the index-name and display-name metadata
- I don't see any copyright lines
- You may add your and Matthew's name and contact information to the AUTHORS file in the root directory of the project
